### PR TITLE
Implement map functions

### PR DESCRIPTION
**Pull request by acquamarin**
1. Implement the following map functions:
a. MAP_EXTRACT: Return a list containing the value for a given key or an empty list if the key is not contained in the map. The type of the key provided in the second parameter must match the type of the map’s keys else an error is returned.	
b.CARDINALITY: Return the size of the map (or the number of entries in the map).	
c. MAP_KEYS: Return a list of all keys in the map.	
d. MAP_VALUES: Return a list of all values in the map.	
2. Fixed list parsing bug.

### CODE REVIEW

File: src/function/vector_map_operation.cpp [hunk 1]
```diff
@@ -32,5 +35,73 @@ std::unique_ptr<FunctionBindData> MapVectorOperations::bindFunc(
...
+            *common::MapType::getValueType(&arguments[0]->dataType)));
+    return std::make_unique<FunctionBindData>(
+        common::LogicalType(common::LogicalTypeID::VAR_LIST, std::move(returnListInfo)));
+}
+
+void MapExtractVectorOperations::validateKeyType(std::shared_ptr<binder::Expression> mapExpression,
+    std::shared_ptr<binder::Expression> extractKeyExpression) {
+    auto mapKeyType = common::MapType::getKeyType(&mapExpression->dataType);
+    if (*mapKeyType != extractKeyExpression->dataType) {
+        throw common::RuntimeException("Unmatched map key type and extract key type");
```
Comment (andyfengHKU) [inline src/function/vector_map_operation.cpp:64]: BinderException. Also I would make it a static function. The principal is, if a function is only used in one cpp file, we should consider make it static and only visible inside that cpp file

**Review by andyfengHKU — APPROVED** [review]

File: src/common/types/types.cpp [hunk 2]
```diff
@@ -64,7 +64,7 @@ StructField::StructField(std::string name, std::unique_ptr<LogicalType> type)
 }
 
 bool StructField::operator==(const kuzu::common::StructField& other) const {
-    return name == other.name && *type == *other.type;
+    return *type == *other.type;
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/common/types/types.cpp:67]: yes, i think we should follow the duckdb way of comparing struct.

File: src/function/vector_map_operation.cpp (hunk 1 above)
Comment (acquamarin) [reply src/function/vector_map_operation.cpp:64]: Made it static function

File: src/common/type_utils.cpp [hunk 3]
```diff
@@ -111,102 +111,216 @@ bool TypeUtils::isValueEqual(
...
+        return false;
+    }
+    auto leftStructFields = common::StructVector::getFieldVectors(leftVector);
+    auto rightStructFields = common::StructVector::getFieldVectors(rightVector);
+    for (auto i = 0u; i < leftStructFields.size(); i++) {
+        auto leftStructField = leftStructFields[i];
+        auto rightStructField = rightStructFields[i];
+        switch (leftStructFields[i]->dataType.getPhysicalType()) {
+        case PhysicalTypeID::BOOL: {
+            if (!isValueEqual(reinterpret_cast<uint8_t*>(leftStructField->getData())[leftEntry.pos],
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/common/type_utils.cpp:238]: done

**Pull request merged**